### DISCUSSION TITLE
exploring-dotnet-6-part-2-comparing-webapplicationbuilder-to-the-generic-host/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#14] [Announcements]
# exploring-dotnet-6-part-2-comparing-webapplicationbuilder-to-the-generic-host/

In this post I take a first look at the new WebApplication and WebApplicationBuilder types in .NET 6, and compare them to the Generic Host from .NET 5.

https://andrewlock.net/exploring-dotnet-6-part-2-comparing-webapplicationbuilder-to-the-generic-host/

**Comment by rosdi** [comment]
This is an awesome explanation. It clears a lot of things.

**Comment by Sergey-PrudentDev** [comment]
Really enjoying the detailed explanations so far, though I'm really looking forward to Section 12, which is why I'm really here.

**Comment by sankar1204** [comment]
Nice deep dive. Helps in better understanding of reason behind the new changes introduced .Net core evolution.

**Comment by matteocontrini** [comment]
Is it possible to use the new `WebApplicationBuilder` to replace the generic host in a worker application? Or we should keep using the old way of doing things? Thanks!

**Reply by andrewlock** [reply]
It's _possible_ but it will add a lot of extra things you don't need, and kestrel will run, so I wouldn't recommend it

**Reply by matteocontrini** [reply]
Thanks! That's what wasn't immediately clear to me.

**Comment by cyimxtck** [comment]
I have something that I am working through, which I leveraged what you said above to understand.

My situation is that I have a web application, which merely serves as a dumb terminal.  The API does all the work, essentially.  Now, I am tasked with doing data migrations from the old system; my idea is that I will leverage what is "beneath" the API in terms of Entity Framework, BLL, etc. to push/pull data in the migration effort.

All I need is a service, so I am trying to build that service using a worker process of Host.

Do you have any guidance on specifically building a process for running background/worker services?  I can buy your book and learn, or another resource that you feel is pertinent, or...

This message tells me really nothing, but I am unable to use the service as you've outlined on your site.

System.InvalidOperationException: 'No constructor for type 'Net._6.Identity.Services.LoggedInUserService' can be instantiated using services from the service container and default values.'

**Reply by cyimxtck** [reply]
This is using .NET 6 in a console application with EF 7.

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 09 2022, 08:55, in reply to [Xavier John](https://disqus.com/by/xavier_john/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/14#discussioncomment-6062751)</em>

---

<p>No, WebApplication is not for worker services, it's for aspnetcore only. Workers in .NET 6 are the same as in .NET 5!</p>

[a comment by an automated account is not shown]

**Comment by MarkRobles** [comment]
Great article to understand and the program.cs and startup code between NET Core versions

**Comment by JoaoCampos07** [comment]
Super interesting, clear and very well written. Even thought we already have the .NET 8 we can learn a lot but knowing the history behind the .NET 6 and the code behind it, to see how things evolved. I can not happy enough that .NET team decided to tackle the all-time issue of Microsoft : to complex for beginners, long learning curve. 
I hope this minimalistic way will continue :)

**Comment by JoaoCampos07** [comment]
Hey, Andrew another great article. 

If you allow me just one comment. Just in this part "This is where the ConfigureWebHostDefaults() extension method on IHostBuilder comes in. This method wraps the IHostBuilder in an internal class, GenericWebHostBuilder", i find it a bit unclear. Just when dig into .net core 3.1 code it as clear for me. Because is not the method ConfigureWebHostDefaults() directly that makes the wrap rather another extension method inside "ConfigureWebHost " : 

``` csharp
public static class GenericHostBuilderExtensions
{
	public static IHostBuilder ConfigureWebHostDefaults(this IHostBuilder builder, Action<IWebHostBuilder> configure)
	{
		// this implementation of IHostBuilder is a HostBuilder() object
		//...
		// With this ConfigureWebHost method the IHostBuilder is wrapped in a IWebHostBuilder 
		// Making it seemless to the end client of the api 
		return builder.ConfigureWebHost(webHostBuilder => 
		{
			WebHost.ConfigureWebDefaults(webHostBuilder);

			configure(webHostBuilder);
		});
	}
 }
```
``` csharp

public static class GenericHostWebHostBuilderExtensions
{
	public static IHostBuilder ConfigureWebHost(this IHostBuilder builder, Action<IWebHostBuilder> configure)
	{
		var webhostBuilder = new GenericWebHostBuilder(builder); // and here we can see that indeed the builder is being wrapped in a GenericWebHostBuilder
		configure(webhostBuilder); // making client configuration
		builder.ConfigureServices((context, services) => services.AddHostedService<GenericWebHostService>()); // making normal web hosting configuration
		return builder;
	}
}

```

**Reply by JoaoCampos07** [reply]
I just realized it is second comment on the same article, sorry :/
